src/node_file.cc

```diff
@@ -829,20 +829,46 @@ static void InternalModuleReadJSON(const FunctionCallbackInfo<Value>& args) {
   }
 
   const size_t size = offset - start;
-  if (size == 0 || (
-    size == SearchString(&chars[start], size, "\"name\"") &&
-    size == SearchString(&chars[start], size, "\"main\"") &&
-    size == SearchString(&chars[start], size, "\"exports\"") &&
-    size == SearchString(&chars[start], size, "\"type\""))) {
-    args.GetReturnValue().Set(env->empty_object_string());
-  } else {
-    Local<String> chars_string =
+  char* p = &chars[start];
+  char* pe = &chars[size];
+  char* pos[2];
+  char** ppos = &pos[0];
+
+  while (p < pe) {
+    char c = *p++;
+    if (c == '"') goto quote;
```

Comment:
could you use if statements here to avoid goto?

Reply:
No, for two reasons:

1. Aesthetic. It keeps the level of indent small.
2. Performance. g++ 9.2.1 at -O3 doesn't move the block out of the way. The goto keeps the inner loop at around 80 bytes:

<details>

```
    10ec:       80 f9 22                cmp    $0x22,%cl
    10ef:       74 1f                   je     1110 <main+0x70>
    10f1:       80 f9 5c                cmp    $0x5c,%cl
    10f4:       75 44                   jne    113a <main+0x9a>
    10f6:       48 39 fa                cmp    %rdi,%rdx
    10f9:       73 65                   jae    1160 <main+0xc0>
    10fb:       0f b6 4e 01             movzbl 0x1(%rsi),%ecx
    10ff:       80 f9 22                cmp    $0x22,%cl
    1102:       74 7c                   je     1180 <main+0xe0>
    1104:       48 89 d6                mov    %rdx,%rsi
    1107:       48 8d 56 01             lea    0x1(%rsi),%rdx
    110b:       80 f9 22                cmp    $0x22,%cl
    110e:       75 e1                   jne    10f1 <main+0x51>
    1110:       49 8d 48 08             lea    0x8(%r8),%rcx
    1114:       49 89 10                mov    %rdx,(%r8)
    1117:       4c 39 c9                cmp    %r9,%rcx
    111a:       72 34                   jb     1150 <main+0xb0>
    111c:       48 8b 34 24             mov    (%rsp),%rsi
    1120:       48 89 f1                mov    %rsi,%rcx
    1123:       48 f7 d1                not    %rcx
    1126:       48 03 4c 24 08          add    0x8(%rsp),%rcx
    112b:       48 83 f9 04             cmp    $0x4,%rcx
    112f:       74 5f                   je     1190 <main+0xf0>
    1131:       4d 89 d0                mov    %r10,%r8
    1134:       48 83 f9 07             cmp    $0x7,%rcx
    1138:       74 76                   je     11b0 <main+0x110>
    113a:       48 39 fa                cmp    %rdi,%rdx
    113d:       73 21                   jae    1160 <main+0xc0>
    113f:       0f b6 0a                movzbl (%rdx),%ecx
    1142:       48 89 d6                mov    %rdx,%rsi
    1145:       eb a1                   jmp    10e8 <main+0x48>
```
</details>

(-O3 is kind of disappointing; with -Os it fits in a single cache line.)

I'll add a comment explaining why it's there.